lib/resolvers/project.js

```diff
@@ -45,34 +45,43 @@ module.exports = ({ models }) => ({ action, data, id, meta }, transaction) => {
       .orderBy('createdAt', 'desc')
       .first()
-      .then(version => version.$query(transaction).patchAndFetch({ status: 'granted' }))
-      .then(version => {
-        let months = get(version, 'data.duration.months', 0);
-        let years = get(version, 'data.duration.years', 5);
-        if (months > 12) {
-          months = 0;
-        }
-        if (years >= 5) {
-          years = 5;
-          months = 0;
-        }
-
-        const project = Project.query(transaction).findById(id).then(project => project);
-        return Promise.all([project, generateLicenceNumber(Project, transaction, 'project', project.schemaVersion)])
-          .then(([project, licenceNumber]) => {
-            const start = project.issueDate ? moment(project.issueDate) : moment();
-            const issueDate = start.toISOString();
-            const expiryDate = start.add({ months, years }).toISOString();
-            return project.$query(transaction).patchAndFetch({
-              status: 'active',
-              issueDate,
-              expiryDate,
-              title: version.data.title,
-              licenceNumber: project.licenceNumber || licenceNumber
-            });
+      .then(grantedVersion => {
+        return ProjectVersion.query(transaction)
+          .where({ projectId: id, status: 'submitted' })
+          .orderBy('createdAt', 'desc')
+          .first()
+          .then(version => version.$query(transaction).patchAndFetch({ status: 'granted' }))
+          .then(version => {
+            const currentDuration = get(grantedVersion, 'data.duration');
+            const newDuration = get(version, 'data.duration');
+            const project = Project.query(transaction).findById(id).then(project => project);
+            return Promise.all([project, generateLicenceNumber(Project, transaction, 'project', project.schemaVersion)])
+              .then(([project, licenceNumber]) => {
+                const start = project.issueDate ? moment(project.issueDate) : moment();
+                const issueDate = start.toISOString();
+                const patch = {
+                  status: 'active',
+                  issueDate,
+                  title: version.data.title,
+                  licenceNumber: project.licenceNumber || licenceNumber
+                };
+                if (!isEqual(currentDuration, newDuration) || newDuration === undefined) {
+                  let months = get(version, 'data.duration.months', 0);
+                  let years = get(version, 'data.duration.years', 5);
+                  if (months > 12) {
+                    months = 0;
+                  }
+                  if (years >= 5) {
+                    years = 5;
+                    months = 0;
+                  }
+                  patch.expiryDate = start.add({ months, years }).toISOString();
+                }
+                return project.$query(transaction).patchAndFetch(patch);
+              });
```

Comment:
There's _a lot_ going on here now. I think there might be some value in pulling some of the transactions out into named functions - e.g. `fetchLatestVersion` etc. to make it a bit clearer.

Reply:
fair.